junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ResourceLock.java

```diff
@@ -43,12 +45,18 @@ default void close() {
 		release();
 	}
 
+	SortedSet<ExclusiveResource> getResources();
+
 	/**
 	 * {@return whether the given lock is compatible with this lock}
 	 * @param other the other lock to check for compatibility
 	 */
 	default boolean isCompatible(ResourceLock other) {
-		return this instanceof NopLock || other instanceof NopLock;
+		if (this.getResources().isEmpty() || other.getResources().isEmpty()) {
+			return true;
+		}
+		return other.getResources().stream() //
+				.allMatch(it -> this.getResources().contains(it)
+						|| ExclusiveResource.COMPARATOR.compare(this.getResources().last(), it) < 0);
```

Comment:
🧐 I would extract the `this.getResources().last()` to a local variable outside the stream.